File: Python/Ejercicio9.py
def merge(arr, l, m, r):
    arr1 = m - l + 1
    arr2 = r - m

    tempL = [0] * arr1
    tempR = [0] * arr2

    for j in range(0, arr1):
        tempL[j] = arr[l + j]

    for k in range(0, arr2):
        tempR[k] = arr[m + 1 + k]

    i = 0
    j = 0
    k = l

    while i < arr1 and j < arr2:
        if tempL[i] <= tempR[j]:
            arr[k] = tempL[i]
            i += 1
        else:
            arr[k] = tempR[j]
            j += 1
        k += 1

    while i < arr1:
        arr[k] = tempL[i]
        i += 1
        k += 1

    while j < arr2:
        arr[k] = tempR[j]
        j += 1
        k += 1

def mergeSort(arr, l, r):
    if l < r:
        m = l + (r - l)//2

        mergeSort(arr, l, m)
        mergeSort(arr, m + 1, r)
        merge(arr, l, m, r)

def main():
    arr = [32, 17, 12, 6, 11, 5, 45]

    print("=== MERGE SORT EN PYTHON ===")
    print(f"Array inicial: {arr}")

    mergeSort(arr.copy(), 0, len(arr) - 1)

    arrSorted = arr.copy()
    mergeSort(arrSorted, 0, len(arrSorted) - 1)
    print(f"Array final: {arrSorted}")

if __name__ == "__main__":
    main()

File: C#/Ejercicio9.cs
using System;

void MergeSort(int[] arr, int l, int r)
{
    if (l < r) {
        int m = l + (r - l) / 2;

        MergeSort(arr, l, m);
        MergeSort(arr, m + 1, r);
        Merge(arr, l, m, r);
    }
}

void Merge(int[] arr, int l, int m, int r)
{
    int arr1 = m - l + 1;
    int arr2 = r - m;

    int[] tempL = new int[arr1];
    int[] tempR = new int[arr2];

    for (int idx = 0; idx < arr1; idx++) {
        tempL[idx] = arr[l + idx];
    }

    for (int idx = 0; idx < arr2; idx++) {
        tempR[idx] = arr[m + 1 + idx];
    }

    int i = 0;
    int j = 0;
    int k = l;

    while (i < arr1 && j < arr2) {
        if (tempL[i] <= tempR[j]) {
            arr[k] = tempL[i];
            i += 1;
        } else {
            arr[k] = tempR[j];
            j += 1;
        }
        k += 1;
    }

    while (i < arr1) {
        arr[k] = tempL[i];
        i += 1;
        k += 1;
    }

    while (j < arr2) {
        arr[k] = tempR[j];
        j += 1;
        k += 1;
    }
}

int[] arr = {64, 34, 25, 12, 22, 11, 90};
Console.WriteLine("=== MERGE SORT EN C# SCRIPT ===");
Console.WriteLine("Array inicial: [" + string.Join(", ", arr) + "]");

MergeSort(arr, 0, arr.Length - 1);

Console.WriteLine("Array final: [" + string.Join(", ", arr) + "]");